frame/multi-account/src/lib.rs

```diff
@@ -0,0 +1,1506 @@
+		///
+		/// If there are enough approvals, then dispatch the call.
+		///
+		/// If one approving account is removed from the signatories of the multi account, it does
+		/// no longer count as a valid approval for the multisig transaction. Removing the initiator/
+		/// owner of a multisig transaction from the signatories will not cancel the multisig transaction.
+		/// It can still be dispatched, as long as the threshold is met by the current signatories.
+		/// Approvals fail if the multi account was removed.
+		///
+		/// Payment: `MultisigDepositBase` will be reserved if this is the first approval, plus
+		/// `threshold` times `MultisigDepositFactor`. It is returned once this dispatch happens or
+		/// is cancelled.
+		///
+		/// The dispatch origin for this call must be _Signed_ and must be one of the signatories.
+		///
+		/// - `multi_account_id`: The account ID of the multi account that was created before.
+		/// - `maybe_timepoint`: If this is the first approval, then this must be `None`. If it is
+		/// not the first approval, then it must be `Some`, with the timepoint (block number and
+		/// transaction index) of the first approval transaction.
+		/// - `call`: The call to be executed.
+		///
+		/// NOTE: Unless this is the final approval, you will generally want to use
+		/// `approve` instead, since it only requires a hash of the call.
+		///
+		/// Result is equivalent to the dispatched result if `threshold` is exactly `1`. Otherwise
+		/// on success, result is `Ok` and the result from the interior call, if it was executed,
+		/// may be found in the deposited `MultisigExecuted` event.
+		///
+		/// # <weight>
+		/// - `O(S + Z + Call)`.
+		/// - Up to one balance-reserve or unreserve operation.
+		/// - One passthrough operation, one insert, both `O(S)` where `S` is the number of
+		///   signatories. `S` is capped by `MaxSignatories`, with weight being proportional.
+		/// - One call encode & hash, both of complexity `O(Z)` where `Z` is tx-len.
+		/// - One encode & hash, both of complexity `O(S)`.
+		/// - Up to two binary searches and inserts (`O(logS + S)`).
+		/// - I/O: 1 read `O(S)`, up to 1 mutate `O(S)`. Up to one remove.
+		/// - One event.
+		/// - The weight of the `call`.
+		/// - Storage: inserts one item, value size bounded by `MaxSignatories`, with a
+		///   deposit taken for its lifetime of
+		///   `MultisigDepositBase + threshold * MultisigDepositFactor`.
+		/// # </weight>
+		#[weight = <MultiPassthrough<<T as Trait>::Call, T::AccountId, Option<Timepoint<T::BlockNumber>>>>::new()]
+		fn call(origin,
+			multi_account_id: T::AccountId,
+			maybe_timepoint: Option<Timepoint<T::BlockNumber>>,
+			call: Box<<T as Trait>::Call>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			if let Some(multi_account) = <MultiAccounts<T>>::get(&multi_account_id) {
+				// ensure that the origin is a signatory of the multi account
+				if let Err(_) = multi_account.signatories.binary_search(&who) {
```

Comment:
User should provide the index in the vector where their account is, and you search a single element to see they are there.

So we change a binary search to a single value lookup.

Reply:
Not sure I agree, that makes the interface much harder for users. This is implemented the same as https://github.com/paritytech/substrate/blob/master/frame/utility/src/lib.rs#L340 right now.

Reply:
It will probably get changed then. In general, its not about "harder for users to use". The JS API can take care of these things where needed. If we can reduce on-chain computation, this means we can reduce the fees we charge to users and increase number of txs included in a single block. This is both much more important. (imo)

Reply:
To be honest it is probably not that bad to keep the same, but I would guess a round of runtime optimization would change it to the other pattern anyway.

Reply:
Makes sense. Happy to change it if this becomes a blocker for getting this merged. Thanks!